### PR TITLE
Variable mappers feedback

### PR DESCRIPTION
**Pull request by johnnycopes**
Explanations incoming

### CODE REVIEW

File: utils/variableMappers.ts [hunk 1]
```diff
@@ -1,25 +1,68 @@
-export type DataValue = string | number | DataObject | DataArray;
+export type DataValue = string | number | DataObject | DataObject[];
 export type DataObject = { [key: string]: DataValue };
-export type DataArray = DataObject[];
```
Comment (johnnycopes) [inline utils/variableMappers.ts:3]: I removed the `DataArray` type and instead wrote `DataObject[]` throughout this file. My take in this particular case was that the additional type definition wasn't worth the mental overhead when TS already has a very readable syntax for arrays. Subjective, but I prefer this approach. 

I would argue that if you decided to keep the type, though, a more accurate name would be `DataObjectArray`

File: utils/variableMappers.ts [hunk 2]
```diff
@@ -1,25 +1,68 @@
...
+
+export function deepTransformKeys(
+  input: DataValue,
+  convertedCase: "camel" | "snake" = "camel"
+): DataValue {
```
Comment (johnnycopes) [inline utils/variableMappers.ts:7]: The return type of this function was the contents of `DataValue`, so I changed it. I imagine you meant for it to return the type rather than `string | number | DataObject | DataObject[]`

File: utils/variableMappers.ts [hunk 3]
```diff
@@ -1,25 +1,68 @@
-export type DataValue = string | number | DataObject | DataArray;
+export type DataValue = string | number | DataObject | DataObject[];
 export type DataObject = { [key: string]: DataValue };
-export type DataArray = DataObject[];
+
+export function deepTransformKeys(
```
Comment (johnnycopes) [inline utils/variableMappers.ts:4]: This is the only function in this file being used outside of this file, so I moved it to the top. Generally, it's useful to keep public functions/methods above private funcitons/methods. The idea is that when reading the code as a consumer, you can easily find the function(s) you're actually calling and then delve into the details of how it works if needed

File: utils/variableMappers.ts [hunk 4]
```diff
@@ -29,7 +72,7 @@ export function toCamelCase(str: string): string {
   return converted;
 }
 
-export function toSnakeCase(str: string): string {
+function toSnakeCase(str: string): string {
```
Comment (johnnycopes) [inline utils/variableMappers.ts:75]: This isn't referenced outside this file so I removed the `export` keyword

File: utils/variableMappers.ts [hunk 5]
```diff
@@ -1,25 +1,68 @@
...
-  Object.entries(snakeToCamelSpecialCases).map(([key, value]) => [value, key])
+const clientKeysRecord: KeysRecord = Object.fromEntries(
+  Object.entries(apiKeysRecord).map(([key, value]) => [value, key])
 );
 
 const camelCaseResultsCache: Record<string, string> = {};
 const snakeCaseResultsCache: Record<string, string> = {};
 
-export function toCamelCase(str: string): string {
+function toCamelCase(str: string): string {
```
Comment (johnnycopes) [inline utils/variableMappers.ts:65]: This isn't referenced outside this file so I removed the `export` keyword

File: utils/variableMappers.ts [hunk 6]
```diff
@@ -1,25 +1,68 @@
...
+
+const apiKeysRecord: KeysRecord = {
+  pk: "id",
 };
 
-const camelToSnakeSpecialCases = Object.fromEntries(
-  Object.entries(snakeToCamelSpecialCases).map(([key, value]) => [value, key])
+const clientKeysRecord: KeysRecord = Object.fromEntries(
+  Object.entries(apiKeysRecord).map(([key, value]) => [value, key])
 );
```
Comment (johnnycopes) [inline utils/variableMappers.ts:60]: 1. These names are tightly coupled to the current mapping implementation. What if next week it turns out you realize it'd actually be better to convert between `kebab-case` and `UpperCamelCase`? The names would no longer make sense -- naming them after their purpose improves their longevity
2. For maximum consistency, I think a local `KeysRecord` type for both records makes it very clear what's happening. The type inference from the logic that swaps the keys/values is essentially the same thing, but not precisely
3. Nifty way to swap the keys/values. I wasn't familiar with `Object.fromEntries`

File: utils/variableMappers.ts [hunk 7]
```diff
@@ -41,46 +84,3 @@ export function toSnakeCase(str: string): string {
...
-
-      if (typeof val === 'object') {
-        newObj[newKey] = transformObjectOrArrayKeys(val, convertedCase);
-      } else {
-        newObj[newKey] = val;
-      }
-    }
-    return newObj;
-  }
-  return input;
```
Comment (johnnycopes) [inline utils/variableMappers.ts:75]: This line is impossible to reach based on your typing of `input`: it will either be a DataObject[] and enter the first block or a DataObject[] and enter the second. Thus, it can be removed

File: utils/variableMappers.ts [hunk 8]
```diff
@@ -1,25 +1,68 @@
...
+        : clientKeysRecord[key] || toSnakeCase(key);
+
+    if (typeof val === "object") {
+      newObj[newKey] = transformObjectOrArrayKeys(val, convertedCase);
+    } else {
+      newObj[newKey] = val;
+    }
+  }
+  return newObj;
+}
```
Comment (johnnycopes) [inline utils/variableMappers.ts:43]: I think there's more that could be done to refactor this, but I don't feel confident in trying anything further because there's no test coverage. I think it could be a valuable to add Jest to the project so we can write some tests, but at a minimum you sharing some test data with me could also work.

Things I notice:

1. I can see why it's useful to make this function recursive, but it both being recursive and manage both objects and arrays makes it complicated. I wonder if there's a way to pull out a function that only handles objects
2. Like (I think) we discussed, there's usually a more type-safe approach to typing rather than using the `as` keyword, which is basically opting out of TS' help

**Review by johnnycopes — COMMENTED** [review]
This generally looks very good. I can tell you've put in a lot of time with JS! Left some suggesitons throughout
